pkg/virt-controller/watch/clone/clone.go

```diff
@@ -0,0 +1,375 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2022 Red Hat, Inc.
+ *
+ */
+
+package clone
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/utils/pointer"
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	clonev1alpha1 "kubevirt.io/api/clone/v1alpha1"
+	k6sv1 "kubevirt.io/api/core/v1"
+	snapshotv1alpha1 "kubevirt.io/api/snapshot/v1alpha1"
+	"kubevirt.io/client-go/log"
+)
+
+type cloneSourceType string
+
+const (
+	sourceTypeVM cloneSourceType = "VirtualMachine"
+)
+
+type cloneTargetType string
+
+const (
+	targetTypeVM cloneTargetType = "VirtualMachine"
+	defaultType  cloneTargetType = targetTypeVM
+)
+
+type syncInfoType struct {
+	err             error
+	snapshotName    string
+	snapshotReady   bool
+	restoreName     string
+	restoreReady    bool
+	targetVMName    string
+	targetVMCreated bool
+
+	isCloneFailing bool
+	failEvent      Event
+	failReason     string
+
+	reenqueueInfo reenqueueInfo
+}
+type reenqueueInfo struct {
+	reenqueueReason string
+}
+
+func (ctrl *VMCloneController) execute(key string) (error, reenqueueInfo) {
+	var syncInfo syncInfoType
+	doNotReenqueue := reenqueueInfo{}
+	logger := log.Log
+
+	obj, cloneExists, err := ctrl.vmCloneInformer.GetStore().GetByKey(key)
+	if err != nil {
+		return err, doNotReenqueue
+	}
+
+	var vmClone *clonev1alpha1.VirtualMachineClone
+	if cloneExists {
+		vmClone = obj.(*clonev1alpha1.VirtualMachineClone)
+		logger = logger.Object(vmClone)
+	} else {
+		return nil, doNotReenqueue
+	}
+
+	var syncErr error
+
+	sourceInfo := vmClone.Spec.Source
+	switch cloneSourceType(sourceInfo.Kind) {
+	case sourceTypeVM:
+		vmKey := getKey(sourceInfo.Name, vmClone.Namespace)
+		obj, vmExists, err := ctrl.vmInformer.GetStore().GetByKey(vmKey)
+		if err != nil {
+			return fmt.Errorf("error getting VM %s in namespace %s from cache: %v", sourceInfo.Name, vmClone.Namespace, err), doNotReenqueue
+		}
+		if !vmExists {
+			err = ctrl.updateStatus(vmClone, syncInfoType{
+				isCloneFailing: true,
+				failEvent:      SnapshotNotCreated,
+				failReason:     fmt.Sprintf("VirtualMachine %s does not exist in namespace %s", vmClone.Spec.Source.Name, vmClone.Namespace),
+			})
+
+			if err != nil {
+				log.Log.Errorf("updating status when source vm does not exist failed: %v", err)
+			}
+
+			return fmt.Errorf("VM %s in namespace %s does not exist", sourceInfo.Name, vmClone.Namespace), doNotReenqueue
+		}
+		sourceVM := obj.(*k6sv1.VirtualMachine)
+
+		syncInfo = ctrl.syncSourceVM(key, sourceVM, vmClone)
+		syncErr = syncInfo.err
+	default:
+		return fmt.Errorf("clone %s is defined with an unknown source type %s", vmClone.Name, sourceInfo.Kind), doNotReenqueue
+	}
+
+	err = ctrl.updateStatus(vmClone, syncInfo)
```

Comment:
The idea of having sync() return a struct that is passed to updateStatus() is rather novel I think?  Not sure we do this anywhere else?  I think I like it.  But it is different.  However I can't think of any reason to not do it right now.

Reply:
Actually, I wasn't familiar with this style either, but I like it too. During my work I've tried to see how other components tackle this problem and found [this usage](https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-controller/watch/vm.go#L316) in virt-controller:

```go
syncErr, err = c.sync(vm, vmi, key, dataVolumes)

// ... some other stuff ...

err = c.updateStatus(vm, vmi, syncErr)
```

What I did in the clone API is a generalization of that approach - provide a whole struct with info about sync status rather than just the sync error.